split out functionalization codegen to use view_copy operators

bdhirsh opened the pull request:
This PR splits the functionalization codegen into 2 pieces:

(1) Vanilla functionalization will now always turn view ops into "view_copy" ops.

(2) For functorch to "reapply views underneath the pass", I added a new dispatch key, "FunctionalizeAddBackViews". I codegen a kernel to that key for every view_copy operator that just calls back into the view op. All other ops get a fallthrough kernel.

Also - the codegen will now unconditionally register CompositeImplicitAutograd kernels directly to the functionalization keys, so we "always decompose" before hitting functionalization. Otherwise, we might break and accidentally send "view" calls to the backend, if we decompose an op into a view underneath the functionalization pass.

The important changes are in `gen.py` and `gen_functionalization_type.py` - most of the other changes are just plumbing `{view}_copy` everywhere. I also updated `test_functionalization.py`, and added expecttests for the "add back views" case.

One thing about the `AddBackViews` key - right now, I add it into the TLS include set. The other option would be to try to add it directly to the tensors, but that's kind of hard: putting it on the `FunctionalTensorWrapper` doesn't help, because the functionalization pass will unwrap when it calls back into the dispatcher, and run on the "inner tensor" (maybe we could modify the inner tensor's keyset and add the `AddBackViews` key when functionalization happens, instead?)

I also have an accompanying functorch change here: https://github.com/pytorch/functorch/pull/678

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75527 [prototype] integrate functionalization <> LTC torchscript backend
* #75913 functionalization: introduce a "zero()" aten op
* #75819 functionalization: avoid some unnecessary view_copy calls
* #75717 fix unfold for meta tensors
* #75716 teach ivalue about List[Optional[Tensor]], fix fallbacks
* #75818 fix out= op handling for functionalization
* **#75302 split out functionalization codegen to use view_copy operators**

Differential Revision: [D35419652](https://our.internmc.facebook.com/intern/diff/D35419652)

bdhirsh committed:
split out functionalization codegen to use view_copy operators

[ghstack-poisoned]

bdhirsh committed:
Update on "split out functionalization codegen to use view_copy operators"




[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35419652).

bdhirsh committed:
Update on "split out functionalization codegen to use view_copy operators"


Differential Revision: [D35419652](https://our.internmc.facebook.com/intern/diff/D35419652)

[ghstack-poisoned]

bdhirsh committed:
Update on "split out functionalization codegen to use view_copy operators"


Differential Revision: [D35419652](https://our.internmc.facebook.com/intern/diff/D35419652)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35419652).

bdhirsh committed:
Update on "split out functionalization codegen to use view_copy operators"


Differential Revision: [D35419652](https://our.internmc.facebook.com/intern/diff/D35419652)

[ghstack-poisoned]

bdhirsh committed:
Update on "split out functionalization codegen to use view_copy operators"


Differential Revision: [D35419652](https://our.internmc.facebook.com/intern/diff/D35419652)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35419652).

bdhirsh committed:
Update on "split out functionalization codegen to use view_copy operators"


Differential Revision: [D35419652](https://our.internmc.facebook.com/intern/diff/D35419652)

[ghstack-poisoned]

bdhirsh committed:
Update on "split out functionalization codegen to use view_copy operators"


Differential Revision: [D35419652](https://our.internmc.facebook.com/intern/diff/D35419652)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_torch_functions_manual.cpp:
```diff
@@ -670,6 +678,7 @@ static PyObject * THPVariable__disable_functionalization(PyObject *self, PyObjec
 {
   HANDLE_TH_ERRORS
   c10::impl::tls_set_dispatch_key_included(at::DispatchKey::Functionalize, false);
+  c10::impl::tls_set_dispatch_key_included(at::DispatchKey::FunctionalizeAddBackViews, false);
```
Isn't this going to prevent clean nesting?
In particular, if the python API should be a context manager, I would expect that it will re-set the state that was there before.
But maybe that's not considered here? If so, should we just forbid enabling it if it is already enabled?

bdhirsh replied:
Yeah you're right, I wasn't thinking too much about the nesting case. Maybe I'll just forbid it now until there's a need.

albanD commented on tools/codegen/context.py:
```diff
@@ -81,15 +82,6 @@ def wrapper(f: F, backend_index: BackendIndex) -> T:
 
 # Convenience decorator for functions that explicitly take in a BackendIndex,
 # instead of indirectly taking one in as a closure
-def with_native_function_and_selector_and_index(
```
The comment above should go as well no?

bdhirsh replied:
The comment also applies to the wrapper below the one that I deleted (technically it takes a dict of BackendIndices, I can fix that)

albanD commented on tools/codegen/gen.py:
```diff
@@ -1651,7 +1669,13 @@ def gen_op_headers(g: Union[NativeFunction, NativeFunctionsViewGroup]) -> List[s
+            'ops_headers',
+            'func_definitions',
+            'func_registrations',
+            'func_add_back_views_definitions',
```
Do we really need to shard since this is only for view functions?

bdhirsh replied:
yeah this is definitely overkill. Let me try to unwind it.

albanD replied:
Ho the out ops are there? Then yes makes sense.

albanD commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -318,13 +325,13 @@ def emit_inplace_functionalization_body(
         functional_call_str = """\
             auto tmp_intermediate = at::_ops::to_other::call(src_, self_, non_blocking, false, c10::nullopt);
-            tmp_output = at::_ops::expand_as::call(tmp_intermediate, self_);"""
+            tmp_output = at::_ops::expand_copy::call(tmp_intermediate, self_.sizes(), false);"""
```
We don't have an `expand_as_copy()` ?

bdhirsh replied:
Yeah this was a bit annoying. We don't have `expand_as_copy` because `expand_as` is a `CompositeImplicitAutograd` op, and at least right now, we don't have `_copy()` variants for `CompositeImpliciitAutograd` views.

We could add them... but they're kinda unnecessary for functionalization because we want decompositions to run before the functionalization pass anyway.

Well, they're unnecessary except for examples like the case above, where we want to do special handling for `copy_()` to call a view op that happens to be composite.

One thing I tried to do was originally was to leave the `expand_as` call, and arange for functionalization to run when it decomposes.  I had some trouble trying to get that to work, so this seemed simpler (manually writing the decomposition into `expand_copy`)

albanD replied:
That makes sense. This new version sounds better.

bdhirsh committed:
Update on "split out functionalization codegen to use view_copy operators"

This PR splits the functionalization codegen into 2 pieces:

(1) Vanilla functionalization will now always turn view ops into "view_copy" ops.

(2) For functorch to "reapply views underneath the pass", I added a new dispatch key, "FunctionalizeAddBackViews". I codegen a kernel to that key for every view_copy operator that just calls back into the view op. All other ops get a fallthrough kernel.

Also - the codegen will now unconditionally register CompositeImplicitAutograd kernels directly to the functionalization keys, so we "always decompose" before hitting functionalization. Otherwise, we might break and accidentally send "view" calls to the backend, if we decompose an op into a view underneath the functionalization pass.

The important changes are in `gen.py` and `gen_functionalization_type.py` - most of the other changes are just plumbing `{view}_copy` everywhere. I also updated `test_functionalization.py`, and added expecttests for the "add back views" case.

One thing about the `AddBackViews` key - right now, I add it into the TLS include set. The other option would be to try to add it directly to the tensors, but that's kind of hard: putting it on the `FunctionalTensorWrapper` doesn't help, because the functionalization pass will unwrap when it calls back into the dispatcher, and run on the "inner tensor" (maybe we could modify the inner tensor's keyset and add the `AddBackViews` key when functionalization happens, instead?)

I also have an accompanying functorch change here: https://github.com/pytorch/functorch/pull/678


Differential Revision: [D35419652](https://our.internmc.facebook.com/intern/diff/D35419652)

[ghstack-poisoned]

bdhirsh commented:
@ezyang I think this is ready for review - this is basically everything pre-LTC-integration to let functionalization always emit view copy operators.

bdhirsh committed:
Update on "split out functionalization codegen to use view_copy operators"

This PR splits the functionalization codegen into 2 pieces:

(1) Vanilla functionalization will now always turn view ops into "view_copy" ops.

(2) For functorch to "reapply views underneath the pass", I added a new dispatch key, "FunctionalizeAddBackViews". I codegen a kernel to that key for every view_copy operator that just calls back into the view op. All other ops get a fallthrough kernel.

Also - the codegen will now unconditionally register CompositeImplicitAutograd kernels directly to the functionalization keys, so we "always decompose" before hitting functionalization. Otherwise, we might break and accidentally send "view" calls to the backend, if we decompose an op into a view underneath the functionalization pass.

The important changes are in `gen.py` and `gen_functionalization_type.py` - most of the other changes are just plumbing `{view}_copy` everywhere. I also updated `test_functionalization.py`, and added expecttests for the "add back views" case.

One thing about the `AddBackViews` key - right now, I add it into the TLS include set. The other option would be to try to add it directly to the tensors, but that's kind of hard: putting it on the `FunctionalTensorWrapper` doesn't help, because the functionalization pass will unwrap when it calls back into the dispatcher, and run on the "inner tensor" (maybe we could modify the inner tensor's keyset and add the `AddBackViews` key when functionalization happens, instead?)

I also have an accompanying functorch change here: https://github.com/pytorch/functorch/pull/678


Differential Revision: [D35419652](https://our.internmc.facebook.com/intern/diff/D35419652)

[ghstack-poisoned]

bdhirsh committed:
Update on "split out functionalization codegen to use view_copy operators"

This PR splits the functionalization codegen into 2 pieces:

(1) Vanilla functionalization will now always turn view ops into "view_copy" ops.

(2) For functorch to "reapply views underneath the pass", I added a new dispatch key, "FunctionalizeAddBackViews". I codegen a kernel to that key for every view_copy operator that just calls back into the view op. All other ops get a fallthrough kernel.

Also - the codegen will now unconditionally register CompositeImplicitAutograd kernels directly to the functionalization keys, so we "always decompose" before hitting functionalization. Otherwise, we might break and accidentally send "view" calls to the backend, if we decompose an op into a view underneath the functionalization pass.

The important changes are in `gen.py` and `gen_functionalization_type.py` - most of the other changes are just plumbing `{view}_copy` everywhere. I also updated `test_functionalization.py`, and added expecttests for the "add back views" case.

One thing about the `AddBackViews` key - right now, I add it into the TLS include set. The other option would be to try to add it directly to the tensors, but that's kind of hard: putting it on the `FunctionalTensorWrapper` doesn't help, because the functionalization pass will unwrap when it calls back into the dispatcher, and run on the "inner tensor" (maybe we could modify the inner tensor's keyset and add the `AddBackViews` key when functionalization happens, instead?)

I also have an accompanying functorch change here: https://github.com/pytorch/functorch/pull/678


Differential Revision: [D35419652](https://our.internmc.facebook.com/intern/diff/D35419652)

[ghstack-poisoned]

bdhirsh committed:
Update on "split out functionalization codegen to use view_copy operators"

This PR splits the functionalization codegen into 2 pieces:

(1) Vanilla functionalization will now always turn view ops into "view_copy" ops.

(2) For functorch to "reapply views underneath the pass", I added a new dispatch key, "FunctionalizeAddBackViews". I codegen a kernel to that key for every view_copy operator that just calls back into the view op. All other ops get a fallthrough kernel.

Also - the codegen will now unconditionally register CompositeImplicitAutograd kernels directly to the functionalization keys, so we "always decompose" before hitting functionalization. Otherwise, we might break and accidentally send "view" calls to the backend, if we decompose an op into a view underneath the functionalization pass.

The important changes are in `gen.py` and `gen_functionalization_type.py` - most of the other changes are just plumbing `{view}_copy` everywhere. I also updated `test_functionalization.py`, and added expecttests for the "add back views" case.

One thing about the `AddBackViews` key - right now, I add it into the TLS include set. The other option would be to try to add it directly to the tensors, but that's kind of hard: putting it on the `FunctionalTensorWrapper` doesn't help, because the functionalization pass will unwrap when it calls back into the dispatcher, and run on the "inner tensor" (maybe we could modify the inner tensor's keyset and add the `AddBackViews` key when functionalization happens, instead?)

I also have an accompanying functorch change here: https://github.com/pytorch/functorch/pull/678


Differential Revision: [D35419652](https://our.internmc.facebook.com/intern/diff/D35419652)

[ghstack-poisoned]

ezyang commented on c10/core/DispatchKey.h:
```diff
@@ -347,11 +347,13 @@ enum class DispatchKey : uint16_t {
   // removal,
   // we can consider adding separate keys dedicated to those individual passes.
   // See Note [Functionalization Pass In Core] for details.
+  FunctionalizeAddBackViews,
```
I'm wondering if we actually need a separate key for this. An alternative is to just have some TLS that Functionalize queries when implementing `_copy` operations.

bdhirsh replied:
Yeah I think I should just do this. We'll save the extra key, and it also feels easier to reason about than the extra redispatch (where in functorch, now you need to worry about having two redispatch's "sandwiched" inside of DynamicLayer).

Current plan: we already have a `FuncTorchTLS` construct in core, and `AddBackViews` is currrently only a useful concept in the context of the `functionalize()` transform, so I'm thinking of adding a flag in there for it.

The functionalize view kernels would just check that TLS right before redispatching (to either a `view` or a `view_copy` op), and functorch's `DynamicLayer` logic would be responsible for setting / un-setting that TLS upon entering / exiting the layer.

(We could probably do something really similar for "preserving mutations", which I remember discussions about that being useful to mobile).

ezyang commented on tools/codegen/api/types.py:
```diff
@@ -641,9 +637,9 @@ def inner_call(self) -> str:
         # Some assertions: lambdas are only used for view ops
         assert f.is_view_op
-        assert functional_op.is_view_op
+        assert not functional_op.is_view_op
```
lol

bdhirsh replied:
(i felt the same way lol)

bdhirsh replied:
Context: before this PR, we have view_inplace kernels that would redispatch to the view kernel (`transpose_` to `transpose`).  So the "functional_op" here was `transpose`, and both were view ops.

Now, `f` will be either a view or view_inplace op, and `functional_op` will always be the view_copy op.

ezyang commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -29,6 +32,11 @@
 #     Generates view_copy() composite kernels for all view_copy operators.
 
 
+FunctionalizationTarget = Enum('Target', (
```
'FunctionalizationTarget', not 'Target'

ezyang replied:
Also, Target is intended to be open, so you could add these there. Though, I'd probably try to follow the existing naming convention (which tries to say if you are DEFINITION or DECLARATION in the name) more closely

ezyang commented on tools/codegen/gen.py:
```diff
@@ -1630,18 +1632,34 @@ def gen_op_headers(g: Union[NativeFunction, NativeFunctionsViewGroup]) -> List[s
+                g,
+                backend_indices[DispatchKey.CompositeImplicitAutograd],
+                FunctionalizationTarget.ADD_BACK_VIEWS,
+            ),
```
It's not clear to me why you don't just generate both the addbackviews and regular things all in one go in a `gen_functionalization_registration` call. Save you a lot of typing here.

ezyang replied:
(the thing Target exists for is when you're mapping over a list of functions, but no mapping is going on here...)

ezyang commented:
updated codegen output would be much appreciated

ezyang commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -371,43 +379,68 @@ def emit_inplace_functionalization_body(
-    view_inverse_sig = ViewInverseSignature(f)
-    return view_inverse_sig.decl()
+    view_api_name = g.view.func.name.unambiguous_name()
+    exprs = ', '.join([e.expr for e in translate(view_copy_sig.arguments(), view_sig.arguments())])
```
Sometime we should add some APIs that make doing this kind of "call-to-call" less wordy haha

ezyang commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -29,6 +32,11 @@
 #     Generates view_copy() composite kernels for all view_copy operators.
```
This comment needs a nice, long update

bdhirsh replied:
yeah... it's actually up-to-date again now that I removed all of the AddBackView kernels

ezyang commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -444,41 +478,78 @@ def emit_registration_helper(f: NativeFunction, *, is_view: bool) -> str:
+    funcTarget: FunctionalizationTarget,
 ) -> List[str]:
-    if not needs_functionalization(selector, g):
+    if not needs_functionalization(selector, g, Target.ANONYMOUS_DEFINITION, funcTarget):
```
how come this one is anonymous?

ezyang commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -444,41 +478,78 @@ def emit_registration_helper(f: NativeFunction, *, is_view: bool) -> str:
-        f = g
-        assert modifies_arguments(f)
-        return [emit_inplace_functionalization_body(f, functional_op)]
+    if funcTarget == FunctionalizationTarget.FUNCTIONALIZATION:
```
it's more idiomatic to do these conditions as

```
if funcTarget is FunctionalizationTarget.FUNCTIONALIZATION:
  ...
elif funcTarget is FunctionalizationTarget.ADD_BACK_VIEWS:
  ...
else:
  assert_never(funcTarget)
```

you then get compile time exhaustiveness checking

bdhirsh replied:
ah yep - this will go away though if I'm using TLS to keep the views around instead of an extra dispatch key / set of kernels to generate.

ezyang commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -371,43 +379,68 @@ def emit_inplace_functionalization_body(
 
-# TODO: later, I'll generate separate kernels for "Functionalize", and "AddBackViews"
-# That will probably be an enum, that this function will take in to know which ops get kernels.
 def needs_functionalization(
```
as you are now unconditionally calling gen_ops_header, it doesn't seem to me like there is any good reason to have factored out the needs functionalization boolean test from the actual codegen. Just do the eligibility test in the codegen; it's not like you're saving any work splitting it up this way.

ezyang commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -444,41 +478,78 @@ def emit_registration_helper(f: NativeFunction, *, is_view: bool) -> str:
+        if isinstance(g, NativeFunctionsViewGroup):
+            # Case 1: emit view -> view_copy kernels for the functionalization pass
+            view_defs = []
+            if not g.view.has_composite_implicit_autograd_kernel:
```
I understand this is from preexisting code, but the test for `g.view.has_composite_implicit_autograd_kernel` seems very obscure. All you're really doing is trying to distinguish between composite views and non-composite views. Is there any reason to put composite views in NativeFunctionsViewGroup? And if so, it seems like it would be better to have this be a property on NativeFunctionsViewGroup itself that can then be documented on what its semantic meaning. Directly referencing `has_composite_implicit_autograd_kernel` feels super low level for the logic here.

bdhirsh replied:
The code here allows for the possibility of a view op being `CompositeImplicitAutograd` but its view_inplace variant to not be (or vice versa).

I just checked though and as per your comment further down: we don't actually have an ops like that today. So maybe I should just enforce "composite-ness" between view and view_inplace ops and simplify this code (make compositeness a property of the group).

ezyang replied:
Yeah, if you can enforce a more restrictive check, defo go for it.

ezyang commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -444,41 +478,78 @@ def emit_registration_helper(f: NativeFunction, *, is_view: bool) -> str:
+                # if the view is not composite (implicit autograd)
+                assert g.view_copy is not None
+                view_defs.append(emit_view_functionalization_body(g.view, g.view_copy))
+            if g.view_inplace is not None and not g.view_inplace.has_composite_implicit_autograd_kernel:
```
is it ever the case that the view inplace has composite implicit autograd, but the view doesn't (or vice versa?)

bdhirsh replied:
(commented above - as of today, nope)

ezyang commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -444,41 +478,78 @@ def emit_registration_helper(f: NativeFunction, *, is_view: bool) -> str:
-            view_str.append(emit_registration_helper(g.view_inplace, is_view=True))
-        return view_str
+        # we always want to register composite kernels
+        if g.view.has_composite_implicit_autograd_kernel:
```
The control flow here matches exactly the codegen logic below, is that right? This is exactly what Target was about; rather than having to keep the control flow thicket in sync, instead you just reuse the control flow and on the inside you then case on the target, so you only have to say the control flow once.

ezyang reviewed (approved):
This looks functionally correct but I do think it can be cleaner.

bdhirsh committed:
Update on "split out functionalization codegen to use view_copy operators"

This PR splits the functionalization codegen into 2 pieces:

(1) Vanilla functionalization will now always turn view ops into "view_copy" ops.

(2) For functorch to "reapply views underneath the pass", I added a new dispatch key, "FunctionalizeAddBackViews". I codegen a kernel to that key for every view_copy operator that just calls back into the view op. All other ops get a fallthrough kernel.

Also - the codegen will now unconditionally register CompositeImplicitAutograd kernels directly to the functionalization keys, so we "always decompose" before hitting functionalization. Otherwise, we might break and accidentally send "view" calls to the backend, if we decompose an op into a view underneath the functionalization pass.

The important changes are in `gen.py` and `gen_functionalization_type.py` - most of the other changes are just plumbing `{view}_copy` everywhere. I also updated `test_functionalization.py`, and added expecttests for the "add back views" case.

One thing about the `AddBackViews` key - right now, I add it into the TLS include set. The other option would be to try to add it directly to the tensors, but that's kind of hard: putting it on the `FunctionalTensorWrapper` doesn't help, because the functionalization pass will unwrap when it calls back into the dispatcher, and run on the "inner tensor" (maybe we could modify the inner tensor's keyset and add the `AddBackViews` key when functionalization happens, instead?)

I also have an accompanying functorch change here: https://github.com/pytorch/functorch/pull/678


Differential Revision: [D35419652](https://our.internmc.facebook.com/intern/diff/D35419652)

[ghstack-poisoned]

bdhirsh committed:
Update on "split out functionalization codegen to use view_copy operators"

This PR splits the functionalization codegen into 2 pieces:

(1) Vanilla functionalization will now always turn view ops into "view_copy" ops.

(2) For functorch to "reapply views underneath the pass", I added a new dispatch key, "FunctionalizeAddBackViews". I codegen a kernel to that key for every view_copy operator that just calls back into the view op. All other ops get a fallthrough kernel.

Also - the codegen will now unconditionally register CompositeImplicitAutograd kernels directly to the functionalization keys, so we "always decompose" before hitting functionalization. Otherwise, we might break and accidentally send "view" calls to the backend, if we decompose an op into a view underneath the functionalization pass.

The important changes are in `gen.py` and `gen_functionalization_type.py` - most of the other changes are just plumbing `{view}_copy` everywhere. I also updated `test_functionalization.py`, and added expecttests for the "add back views" case.

One thing about the `AddBackViews` key - right now, I add it into the TLS include set. The other option would be to try to add it directly to the tensors, but that's kind of hard: putting it on the `FunctionalTensorWrapper` doesn't help, because the functionalization pass will unwrap when it calls back into the dispatcher, and run on the "inner tensor" (maybe we could modify the inner tensor's keyset and add the `AddBackViews` key when functionalization happens, instead?)

I also have an accompanying functorch change here: https://github.com/pytorch/functorch/pull/678


Differential Revision: [D35419652](https://our.internmc.facebook.com/intern/diff/D35419652)

[ghstack-poisoned]

bdhirsh commented:
Pushed another round of updates. List of changes:

(1) Removed the `FunctionalizeAddBackViews` dispatch key, and used TLS instead, inside of the functionalization kernels, to decide whether or not I redispatch to a view or view_copy operator.

The main annoying thing here is for the view_inverse lambdas. I didn't want to have to manually write out twice as many "view_inverse" functions, so instead I updated them all to plumb through a `reapply_views` argument, which they use to decide whether to call the underlying view or view_copy op. I had to update the `at::sum_to()` helper to also take in an optional `always_return_non_views` flag (thought it was cleaner than duplicating the function just for functionalization).

(2) removed `needs_functionalization()` since it was redundant, and cleaned up the "when to generate a kernel" logic a bit by making `composite` a property of the `NativeFunctionsViewGroup`. Turns out there is one view group that doesn't have matching composite info: `narrow` is `CompositeImplicitAutograd`, but `narrow_copy` is Composite

(3) I ended up plumbing `NativeFunctionsViewGroup` further into the codegen layer, into `FunctionalizationLambda`. This let me remove a few assertions, and also fixed an issue where in the AddBackView case for `transpose_()`, I needed to be able to have access to the name + schema of the view op `transpose()` (before I only had the view_inplace and view_copy lying around).

bdhirsh committed:
Update on "split out functionalization codegen to use view_copy operators"

This PR splits the functionalization codegen into 2 pieces:

(1) Vanilla functionalization will now always turn view ops into "view_copy" ops.

(2) For functorch to "reapply views underneath the pass", I added a new dispatch key, "FunctionalizeAddBackViews". I codegen a kernel to that key for every view_copy operator that just calls back into the view op. All other ops get a fallthrough kernel.

Also - the codegen will now unconditionally register CompositeImplicitAutograd kernels directly to the functionalization keys, so we "always decompose" before hitting functionalization. Otherwise, we might break and accidentally send "view" calls to the backend, if we decompose an op into a view underneath the functionalization pass.

The important changes are in `gen.py` and `gen_functionalization_type.py` - most of the other changes are just plumbing `{view}_copy` everywhere. I also updated `test_functionalization.py`, and added expecttests for the "add back views" case.

One thing about the `AddBackViews` key - right now, I add it into the TLS include set. The other option would be to try to add it directly to the tensors, but that's kind of hard: putting it on the `FunctionalTensorWrapper` doesn't help, because the functionalization pass will unwrap when it calls back into the dispatcher, and run on the "inner tensor" (maybe we could modify the inner tensor's keyset and add the `AddBackViews` key when functionalization happens, instead?)

I also have an accompanying functorch change here: https://github.com/pytorch/functorch/pull/678


Differential Revision: [D35419652](https://our.internmc.facebook.com/intern/diff/D35419652)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35419652).

ezyang commented:
> I didn't want to have to manually write out twice as many "view_inverse" functions, so instead I updated them all to plumb through a reapply_views argument, which they use to decide whether to call the underlying view or view_copy op.

This seems troublesome, because doesn't this make it impossible to give accurate schema information for the view inverse function? You don't know if it will alias or not without knowing the boolean setting.
